### PR TITLE
Fix Rails and Style Autocorrect Rubocop Errors

### PR DESCRIPTION
**Pull request by macanudo527**
#### What? Why?

- Part of #11253

This fixes 3 more rubocop rules that are violated in our codebase. It was generated automatically with the [rubocop-autocorrect.sh](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/script/rubocop-autocorrect.sh) script.

The correction also seems to fix a subscription bug. When no items where available then it looked like all items were available. This is just true for a part of the subscription code though and there may be other parts covering this. The bug may not have been visible at all.

#### What should we test?

* Test placing a subscription.
* Test the edge case of all subscription items becoming unavailable (no stock or removed from the order cycle).
* Don't worry about reproducing the bug, it may have been hidden, but test that it now works as expected: you should receive an email that the order is empty: "Unfortunately, none of products that you ordered were available, so no order has been placed. The original quantities that you requested appear crossed-out below."

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -6,6 +6,14 @@
...
 # versions of RuboCop, may require this file to be generated again.
 
+# Offense count: 1
+# This cop supports safe autocorrection (--autocorrect).
+# Configuration parameters: Max, AllowHeredoc, AllowURI, URISchemes, IgnoreCopDirectives, AllowedPatterns.
+# URISchemes: http, https
+Layout/LineLength:
+  Exclude:
+    - 'spec/system/admin/tag_rules_spec.rb'
+
```
Comment (mkllnk) [inline .rubocop_todo.yml:16]: Do you think that it would be worth re-running rubocop --regenerate-todo before commit so that new violations are not listed in the next commit? It would slow down the process but it would create a better commit history and is automatic.

Or we stop the script if there are new violations so that they can be fixed first.

File: app/services/cap_quantity.rb [hunk 2]
```diff
@@ -42,7 +42,7 @@ def reload_order
   end
 
   def unavailable_stock_lines_for
-    order.line_items.where('variant_id NOT IN (?)', available_variants_for.select(&:id))
+    order.line_items.where.not(variant_id: available_variants_for.select(&:id))
```
Comment (mkllnk) [inline app/services/cap_quantity.rb:45]: This breaks a spec but I haven't found out yet why.

**Review by mkllnk — CHANGES_REQUESTED** [review]

File: app/services/cap_quantity.rb (hunk 2 above)
Comment (mkllnk) [reply app/services/cap_quantity.rb:45]: Here's the difference in generated SQL:
```diff
-(variant_id NOT IN (NULL))
+1=1
```

The result of `available_variants_for.select(&:id)` is an empty array `[]`. So I guess that the difference is that that `1=1` is always true while `variant_id NOT IN (NULL)` is always false. I don't know which one is correct though.

I think, when no variants are available then all variants should be in `unavailable_stock_lines_for`. So your new version is actually fixing a bug. But that case needs to be added to the spec for this service and the currently failing spec needs to be fixed by making the line items available for shipping.

**Review by mkllnk — APPROVED** [review]
I fixed the failing spec and added more spec coverage to make sure that the new version with `where.not` is doing the right thing. This is a good of example of bad code style not just being difficult to read but sometimes also hiding bugs.

File: .rubocop_todo.yml (hunk 1 above)
Comment (macanudo527) [reply .rubocop_todo.yml:16]: I think re-running would be the best option. Like you said earlier, if you start a script, you want it to just run by itself and not stop a minute into it for more user input.

Patching up the lingering errors in a final PR shouldn't be too big of a deal.

File: spec/system/admin/tag_rules_spec.rb [hunk 3]
```diff
@@ -206,41 +206,39 @@
                        from: "enterprise_tag_rules_attributes_4_preferred_matched_" \
                              "shipping_methods_visibility"
       end
-=begin
```
Comment (dacook) [inline spec/system/admin/tag_rules_spec.rb:209]: TIL Ruby has block comments (and that our style guide disallows them).

File: spec/models/order_balance_spec.rb [hunk 4]
```diff
@@ -45,7 +45,9 @@
     end
 
     it 'returns the balance wraped in a Money object' do
-      expect(order_balance.display_amount).to eq(Spree::Money.new(20, currency: ENV['currency']))
+      expect(order_balance.display_amount).to eq(Spree::Money.new(20,
+                                                                  currency: ENV.fetch('currency',
+                                                                                      nil)))
```
Comment (dacook) [inline spec/models/order_balance_spec.rb:50]: I was curious about this rule to [looked it up](https://stackoverflow.com/questions/47985660/rails-difference-between-env-fetch-and-env) to see why. 

> `fetch` allows you to specify the behavior if a key is not found

Which can be to return a default value, or raise an exception.

I guess this is intended to remind the developer that they should consider those options. In fact, I think the latter would be more appropriate in this and the preceding spec. 
I don't think the message makes that option clear:

> Style/FetchEnvVar: Use ENV.fetch('DISCOURSE_URL') or ENV.fetch('DISCOURSE_URL', nil) instead of ENV['DISCOURSE_URL'].

So to me, this just makes the code unnecessarily longer without improving quality. 

But.. I don't think we need to debate it, let's just accept the standard and stick with it ;)

File: spec/services/cap_quantity_spec.rb [hunk 5]
```diff
@@ -83,5 +83,38 @@
...
+        expect {
+          changes = CapQuantity.new.call(order)
+          [line_item1, line_item2, line_item3].each(&:reload)
+        }.to change { line_item1.quantity }.to(0)
+          .and change { line_item2.quantity }.to(0)
+          .and change { line_item3.quantity }.to(0)
+
+        expect(changes[line_item1.id]).to eq 3
+        expect(changes[line_item2.id]).to eq 3
+        expect(changes[line_item3.id]).to eq 3
```
Comment (dacook) [inline spec/services/cap_quantity_spec.rb:100]: I don't understand the purpose of the `changes` attribute. I would expect the "changes" to be `-3` in this case. 
Rather, it seems to be the original quantity of the line item before the cap was applied, which is quite a different meaning (and I'm not sure is useful at all).

🤷 I can see that this is already how the CapQuantity class works, so I guess it's all good.

**Review by dacook — APPROVED** [review]
Looks good. 

A couple of items don't seem quite optimal to me.. but I don't think it's worth following up so mark ready for testing.

**Comment by filipefurtad0** [comment]
Hey again @macanudo527 ,

Four different scenarios were tested; I'm adding a picture of the email received by the customer, on each one:

i) subscription order placement, with no stock restrictions - all as before :heavy_check_mark: 
ii) subscription order placement, with partial stock restriction (ex.: 3 items to order, only 1 available) - order created, with the available stock :heavy_check_mark: 

image

iii) subscription order placement, with total stock restriction (ex.: 3 items to order, 0 available) - no order created :heavy_check_mark: 

image

iv) subscription order placement, with product removed from OC - order not created :heavy_check_mark: 

Same as the test case above; if the product is removed from the order cycle, the message refers stock outage:

image

This looks good, I've spotted no errors: the app seems resilient to stock outages, on what subscriptions are concerned - no bugsnag errors, emails all arriving as expected.

Thanks so much for your effort :muscle:  Merging.
